lib/drb/drb.rb

```diff
@@ -357,27 +358,67 @@ class DRbConnError < DRbError; end
   # For alternative mechanisms, see DRb::TimerIdConv in drb/timeridconv.rb
   # and DRbNameIdConv in sample/name.rb in the full drb distribution.
   class DRbIdConv
+    MUTEX = Mutex.new
+
+    def initialize
+      @id2ref = {}
```

Comment:
Would `ObjectSpace::WeakMap` make sense here? That would keep the entries as long as the value is alive, which seems the desired semantic here.

Reply:
That works only for Ruby 2.7 and higher, since it would require immediates (integer IDs) as keys. That restriction might be ok, since 2.6 did not ship drb as a gem. It would mean this gem is not usable at all on 2.6.

```
$ rvm ruby-2.6.7 do ruby -e 'ObjectSpace::WeakMap.new[self.__id__] = self'
-e:1:in `[]=': cannot define finalizer for Integer (ArgumentError)
	from -e:1:in `<main>'
```

Reply:
This gem specifies >= 2.7 so that seems fine:
https://github.com/ruby/drb/blob/82643ca872c602ec7f6dde7731391be557a281d1/drb.gemspec#L16

Also, WeakIdConv already uses `ObjectSpace::WeakMap`.
I think making `WeakIdConv` the default is probably the safest way.

Reply:
I would suggest that we should also remove the id2ref version in order to start moving away from having any official code that uses it. The WeakMap version will be more reliable and should work on 2.7+ compatible implementations.